feat: add point randomness to sameMessage verification

matthewkeil opened the pull request:
**Ready for review.  Please do not merge until metrics are verified though!!!**

improve bls verification, see #6462 for more context.

Resolves #6462 for SWIG version of the bls bindings and rebuild of the blst-ts library already has fix applied.  Will need to update this code slightly to mimic the work done here.  
https://github.com/ChainSafe/lodestar/compare/mkeil/test-blst-7...mkeil/test-blst-mult-in-verification-2?expand=1

The result from that branch `mkeil/test-blst-mult-in-verification-2` are the ones displayed in the issue above.  Deploying this branch to `feat2` and will post metrics/results to this PR.

matthewkeil committed:
feat: add point randomness to sameMessage verification

matthewkeil committed:
chore: lint

matthewkeil committed:
fix: bls unit test

matthewkeil committed:
feat: account for both point types when applying mult

jeluard commented on packages/beacon-node/src/chain/bls/multithread/jobItem.ts:
```diff
@@ -70,6 +73,42 @@ export function jobItemWorkReq(job: JobQueueItem, format: PointFormat, metrics:
+      if (job.opts.addVerificationRandomness) {
+        const pkPoint = new swigBindings.blst.P1();
+        const sigPoint = new swigBindings.blst.P2();
+        for (let i = 0; i < job.sets.length; i++) {
```
Looks like we can iterate over set directly, bypassing indexes?

matthewkeil replied:
good call!!! copy/pasta. Updated!

jeluard commented on packages/beacon-node/src/chain/bls/multithread/jobItem.ts:
```diff
@@ -70,6 +73,42 @@ export function jobItemWorkReq(job: JobQueueItem, format: PointFormat, metrics:
+          // not show that this is possible
+          const {value} = job.sets[i].publicKey as unknown as blst.PublicKey;
+          let dup: swigBindings.P1;
+          if (typeof (value as swigBindings.P1_Affine).to_jacobian === "function") {
```
Is there anything that could be done on `bls` side to remove:
* the blst specific cast
* the `to_jacobian` check
* the `blst.Signature` usage

Might be worth creating an issue in the `bls` repo?

Also, what does that mean if this is executed in `env` not running native `bls` security-wise? (retorical question for now as there is no such env AFAIK)

matthewkeil replied:
Yes and no.  Short answer is no though.  the mult() functionality is not implemented by the isomorphic bls library because herumi does not support it (see the comment response below for my attempt to implement).  So we need to step down to the blst lib to get the functionality we need to resolve the issue.  The `to_jacobian` check is implemented in bls sort of (the PublicKey class extends the blst PublicKey class) and parts of that code are actually non-dry and using what is in blst.  

It is possible to use bls to get access to that method but it would take a couple hacks because the constructor of PublicKey is private and we would also need to new up a new object that would immediately get thrown away afterwards so for performance reasons I just skipped that step.  

But for completeness this is what it would look like.  I am open to changing though as this is not something i feel strongly about but hopefully this code will all go away in a few weeks when we get the blst rebuild merged.  This was just an acceleration to get the potential security issue cleared up quickly

```ts
const point = (set.publicKey as unknown as blst.PublicKey).value.dup();  // create dup
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
const pk = new bls.PublicKey(point); // new up bls wrapper objects
pkPoint.add((pk as blst.PublicKey).jacobian.mult(randomness)); // create copy of point with `.jacobian`
```

versus 

```ts
let dup: swigBindings.P1;
if (typeof (value as swigBindings.P1_Affine).to_jacobian === "function") {   // either create dup or fresh point
    dup = (value as swigBindings.P1_Affine).to_jacobian();
} else {
    dup = value.dup() as swigBindings.P1;
}
pkPoint.add(dup.mult(randomness));
```

jeluard commented on packages/beacon-node/src/chain/bls/multithread/jobItem.ts:
```diff
@@ -70,6 +73,42 @@ export function jobItemWorkReq(job: JobQueueItem, format: PointFormat, metrics:
+          // not show that this is possible
+          const {value} = job.sets[i].publicKey as unknown as blst.PublicKey;
+          let dup: swigBindings.P1;
+          if (typeof (value as swigBindings.P1_Affine).to_jacobian === "function") {
```
Is there anything that could be done on `bls` side to remove:
* the blst specific cast
* the `to_jacobian` check
* the `blst.Signature` usage

Might be worth creating an issue in the `bls` repo?

Also, what does that mean if this is executed in `env` not running native `bls` security-wise? (retorical question for now as there is no such env AFAIK)

matthewkeil replied:
duplicate. resolving.

jeluard commented on packages/beacon-node/src/chain/bls/interface.ts:
```diff
@@ -22,6 +22,14 @@ export type VerifySignatureOpts = {
   priority?: boolean;
 };
 
+export interface QueuedVerificationOptions extends VerifySignatureOpts {
```
Nitpick but weird that one is spelled `Options` while the other one is `Opts`.

matthewkeil replied:
Nit noted and its a good point!!  I actually thought about this as well but love to fully spell words and didn't want to grow the scope changes.  Updated!

jeluard commented on packages/beacon-node/src/chain/bls/multithread/index.ts:
```diff
@@ -134,9 +136,14 @@ export class BlsMultiThreadWorkerPool implements IBlsVerifier {
     // TODO: Allow to customize implementation
     const implementation = bls.implementation;
+    if (implementation === "herumi") {
+      // mult not implemented by base herumi library
```
Worth creating an issue in `bls` repo?

matthewkeil replied:
Yes very much so!! I mentioned this in passing on one of the calls we did together as a good idea for something that should be done.  An issue will def formalize it.  I did some research and found that there is a `mult` free function included in the `bls-eth-wasm` library but it was not so easily accessible.  It looks like its being exported but was not part of the TS definition so need to dig a bit to figure out how to get a hold of it.  There are a couple of factory functions that initialize the herumi library and it was not clear, without more research, on how to implement.  Should be possible though and potentially when you updated the library the new definitions expose the helper free functions!!!  TBD though.

jeluard commented on packages/beacon-node/src/chain/bls/multithread/utils.ts:
```diff
@@ -17,3 +19,16 @@ export function chunkifyMaximizeChunkSize<T>(arr: T[], minPerChunk: number): T[]
+export function randomBytesNonZero(bytesCount: number): Buffer {
+  const rand = crypto.randomBytes(bytesCount);
+  for (let i = 0; i < bytesCount; i++) {
+    if (rand[0] !== 0) return rand;
```
Why the loop? `i` is not accessed here. Or is it a time-attack protection?

Also, looks like this code comes from `blst` repo. Maybe it should be made available from there?

twoeths replied:
```suggestion
    if (rand[i] !== 0) return rand;
```

matthewkeil replied:
good catch!!! yes... this was copy/pasta from the blst library and i did not notice that bug before.  This should be updated in the base library as well.  I will put up a PR with that fix and export it so we can use it.  For now I have updated this to be functionally correct and once the PR from blst is exported we can remove this and use that one.

matthewkeil commented:
Metrics look pretty good initially.  Metrics show last 6 hours to avoid restart transience.  Added an extra panel for the timer that is surrounding the main thread aggregation where the mult is happening.  The BLS job wait time is way up as is the attestation gossip validation time and wait time.  This was likely (but not completely) due to feat2 having better mesh peering and roughly double the amount of traffic as stable.

UPDATED: I rearranged the gif for easier comparison. In all situations below `stable-mainnet-hzax41` is first and `feat2-mainnet-hzax41` follows.

![Screenshot 2024-03-08 at 2 01 49 PM](https://github.com/ChainSafe/lodestar/assets/18608739/640d9f09-9bca-4131-8d91-cfcb487e6920)
![Screenshot 2024-03-08 at 2 01 34 PM](https://github.com/ChainSafe/lodestar/assets/18608739/a7184a67-2707-4c50-9bc2-e842efee0582)

![Screenshot 2024-03-08 at 2 04 26 PM](https://github.com/ChainSafe/lodestar/assets/18608739/7a095a9f-ce51-4cec-83cb-8e2ccbee1e8e)
![Screenshot 2024-03-08 at 2 03 52 PM](https://github.com/ChainSafe/lodestar/assets/18608739/bddc499e-dbcd-480b-92b8-78e41c207fc4)

![Screenshot 2024-03-08 at 2 04 34 PM](https://github.com/ChainSafe/lodestar/assets/18608739/cd631d88-32a8-48ee-8a78-3a8beb263a22)
![Screenshot 2024-03-08 at 2 04 02 PM](https://github.com/ChainSafe/lodestar/assets/18608739/a6efb849-2af5-4645-93a3-1041f8c8a5ff)

<img width="854" alt="Screenshot 2024-03-08 at 2 13 41 PM" src="https://github.com/ChainSafe/lodestar/assets/18608739/f2930b3b-4906-4e98-b823-b7d366bdfcc3">
<img width="858" alt="Screenshot 2024-03-08 at 2 16 32 PM" src="https://github.com/ChainSafe/lodestar/assets/18608739/e4bc2524-8b48-41cd-b9f8-74c3a4fb55fe">

<img width="844" alt="Screenshot 2024-03-08 at 2 13 48 PM" src="https://github.com/ChainSafe/lodestar/assets/18608739/1330dff4-8472-4d61-be4b-9803bc632d26">
<img width="855" alt="Screenshot 2024-03-08 at 2 16 39 PM" src="https://github.com/ChainSafe/lodestar/assets/18608739/30e23fb5-64e9-4c03-9a27-a1899ccd8455">

<img width="853" alt="Screenshot 2024-03-08 at 2 23 18 PM" src="https://github.com/ChainSafe/lodestar/assets/18608739/1f34f88d-a40b-4aa7-8593-fa121aadf4ef">
<img width="852" alt="Screenshot 2024-03-08 at 2 23 28 PM" src="https://github.com/ChainSafe/lodestar/assets/18608739/4560c964-5e7f-4f06-9d55-fd2e405e0dab">

<img width="851" alt="Screenshot 2024-03-08 at 2 20 30 PM" src="https://github.com/ChainSafe/lodestar/assets/18608739/df24b41a-5d49-4819-940f-17f142a369a2">
<img width="859" alt="Screenshot 2024-03-08 at 2 20 14 PM" src="https://github.com/ChainSafe/lodestar/assets/18608739/7ec82d57-e5b6-4bbb-8848-9846a54f93ee">

matthewkeil committed:
refactor: rename VerifySignatureOptions

matthewkeil committed:
refactor: use for of loop in randomness loop

matthewkeil committed:
fix: bug in randomBytesNonZero

wemeetagain commented:
closing in favor of #6616
